src/DependencyInjection/DI.Specification.Tests/src/DependencyInjectionSpecificationTests.cs

```diff
@@ -545,6 +545,160 @@ public void OpenGenericServicesCanBeResolved()
             Assert.Same(singletonService, genericService.Value);
         }
 
+        [Fact]
```

Comment:
What happens if a singular service is resolved and the generic argument doesn't satisfy its constraint?

Reply:
I wasn't sure about adding that test since there's not an equivalent spec test against resolving a missing service. I can add one, though.

Reply:
The behavior is to return `null`, fyi.

Reply:
Good point, lets add it to `ServiceProviderContainerTests` instead.

Reply:
Hm, am I correct in assuming that it would throw an exception before this change? How hard is it to keep throwing in singular resolution cases?

Reply:
It doesn't have to be part of the spec, failure cases generally aren't. I'm just trying to figure out what would be a better diagnostics experience for customers and if we should change it in this case.

Reply:
When you call `GetRequiredService` and container returns `null` user would get `No service for type '{0}' has been registered.` which would be quite misleading in case of not satisfied constraints.

Reply:
Ah, I see. I had added a failure case, but I'll remove it. I honestly don't know what customers would expect - I generally don't try to resolve single constrained services. It's almost always part of an enumerable, for a chain of responsibility pattern or the like.

Personally, I'd prefer null, as an exception would mean I would need to add some nullo type just to allow a resolution.

Reply:
This might be a difficult one to signal, too, since you're really looking for the case that open generic service descriptors were found that matched the open generic type, but not the type parameter constraints. Partial matches don't really fit the API. It's either `null` or some implementation of `ServiceCallSite`. Even if the message is technically correct, I agree it's not too helpful in this scenario.

Reply:
It's not very difficult I think. Adding a flag to `TryCreateOpenGeneric` to say if it should skip or throw on non-satisfiable constraints and setting it to skip in TryCreateIenumerable.

Reply:
Ah I see it now!

Reply:
I think I'm missing something here - `GetRequiredService` just calls `GetService`, then checks for `null`. How would `CallSiteFactory` know if it should throw or not in the case of a single service? As a consumer, I'd only want to see the specific exception on `GetRequiredService` but I don't see the connection between that method all the way down to `CallSiteFactory`, where the constraint check happens.

Reply:
Ah, I meant throw exception all the time both for `GetService` and `GetRequiredService` like we currently do. It's not super important scenario anyway so let's not change the behaviour.

Reply:
Sounds good! I kept the test in `CallSiteFactory` to confirm returning null, and removed the test from the specs.

Reply:
Sorry if I wasn't clear enough, I'd like us to continue throwing in singular resolution case as we did before.

Reply:
Just to make sure I have this correct:

- `GetService<INonexistentService>()` - returns `null`
- `GetRequiredService<INonexistentService>()` - throws
- `GetService<IFakeGenericService<INonmatchingConstraint>>()` - throws
- `GetRequiredService<IFakeGenericService<INonmatchingConstraint>>()` - throws

For the third case, it needs to check all potential matches before throwing. I can't just let the existing try-catch throw - that was the original problem. It needs to exhaust all options, and skip non-matching ones.

I don't want the existing exception to be thrown, either. It's `ArgumentException` from trying to close a generic type, which is different than "There were services registered, but none whose constraints matched".

This seems like an inconsistency between "regular" services returning null for `GetService` but that one throwing for no services matching the constraint.

In any case, if we throw, it looks like we'd want a more specific exception than what currently gets thrown by the BCL. And I'd only want the more specific exception thrown when concrete types *were* found, but *none* matched.

If no concrete types were found, I'd expect the existing behavior to stand - return null, then let `GetService`/`GetRequiredService` behavior determine exception throwing.

Reply:
This is how I currently see it:

# Single registration - single resolution

Registration:
```
IService<> -> Constrained<>
```

Resolution:
```
IService<Matching> - Success
IService<NonMatching> - Exception
```

# Multiple registrations - single resolution

Registration:
```
IService<> -> NonConstrained<>
IService<> -> Constrained<>
```

Resolution:
```
IService<Matching> - Success
IService<NonMatching> - Exception
```

# Single registration - multiple resolution

Registration:
```
IService<> -> Constrained<>
```

Resolution:
```
IEnumerable<IService<Matching>> - Success 1 results
IEnumerable<IService<NonMatching>> - Success 0 results
```

# Multiple registrations - multiple resolution

Registration:
```
IService<> -> NonConstrained<>
IService<> -> Constrained<>
```

Resolution:
```
IEnumerable<IService<Matching>> - Success 2 results
IEnumerable<IService<NonMatching>> - Success 1 result
```


Same rules apply when IService is dependency of other service being resolved.

Reply:
To summarize:

- Enumerables return empty
- Singular throw

For singular, do you want to expose the `ArgumentException` directly or bubble up something else, like `InvalidOperationException`? Keep the `ArgumentException` as the inner?

Reply:
Having a nice exception is better if it doesn't overcomplicate things too much.

Reply:
And I think this does make sense now - it's more of an edge case to have singular constrained service resolution, and then you get into weird precedence rules if you try to emulate how the compiler chooses the "right" one (like with generic methods and overloads)

Reply:
Added an exception in the singular case, with its own localizable message.